From 27aa4e07743cb8c5545d8813da9aad3788a3290b Mon Sep 17 00:00:00 2001
From: Steve Dickson <SteveD@redhat.com>
Date: Thu, 18 Sep 2014 14:20:42 -0400
Subject: [fs] nfs: reject changes to resvport and sharecache during remount

Message-id: <1411050057-12417-80-git-send-email-steved@redhat.com>
Patchwork-id: 96231
O-Subject: [RHEL7.1 PATCH 79/94] nfs: reject changes to resvport and sharecache during remount
Bugzilla: 1111708
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>

From: Scott Mayhew <smayhew@redhat.com>

Commit c8e47028 made it possible to change resvport/noresvport and
sharecache/nosharecache via a remount operation, neither of which should be
allowed.

Signed-off-by: Scott Mayhew <smayhew@redhat.com>
Fixes: c8e47028 (nfs: Apply NFS_MOUNT_CMP_FLAGMASK to nfs_compare_remount_data)
Cc: stable@vger.kernel.org # 3.16+
Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Upstream Id: 71a6ec8ac587418ceb6b420def1ca44b334c1ff7
Bugzilla: 1111708
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfs/super.c b/fs/nfs/super.c
index bb7762d..02b0b3f 100644
--- a/fs/nfs/super.c
+++ b/fs/nfs/super.c
@@ -2178,7 +2178,7 @@ out_no_address:
  return -EINVAL;
 }
 
-#define NFS_MOUNT_CMP_FLAGMASK ~(NFS_MOUNT_INTR \
+#define NFS_REMOUNT_CMP_FLAGMASK ~(NFS_MOUNT_INTR \
   | NFS_MOUNT_SECURE \
   | NFS_MOUNT_TCP \
   | NFS_MOUNT_VER3 \
@@ -2186,15 +2186,16 @@ out_no_address:
   | NFS_MOUNT_NONLM \
   | NFS_MOUNT_BROKEN_SUID \
   | NFS_MOUNT_STRICTLOCK \
-  | NFS_MOUNT_UNSHARED \
-  | NFS_MOUNT_NORESVPORT \
   | NFS_MOUNT_LEGACY_INTERFACE)
 
+#define NFS_MOUNT_CMP_FLAGMASK (NFS_REMOUNT_CMP_FLAGMASK & \
+  ~(NFS_MOUNT_UNSHARED | NFS_MOUNT_NORESVPORT))
+
 static int
 nfs_compare_remount_data(struct nfs_server *nfss,
     struct nfs_parsed_mount_data *data)
 {
- if ((data->flags ^ nfss->flags) & NFS_MOUNT_CMP_FLAGMASK ||
+ if ((data->flags ^ nfss->flags) & NFS_REMOUNT_CMP_FLAGMASK ||
      data->rsize != nfss->rsize ||
      data->wsize != nfss->wsize ||
      data->version != nfss->nfs_client->rpc_ops->version ||
-- 
1.7.1